postprocessing/comstockpostproc/utils/hpc.py

```diff
@@ -654,43 +655,121 @@ def count_job_errs(job_err_path, errs_dir):
+                # Sort by second-level keys and then by top-level keys
+                for second_key in sorted(flipped_dict.keys()):
+                    f.write(f'{second_key}:\n')
+                    for top_key in sorted(flipped_dict[second_key].keys()):
+                        f.write(f'\t{top_key}:\n')
+                        for error_line in flipped_dict[second_key][top_key]:
+                            f.write(f'\t{error_line}')
+            else:
+                raise ValueError("Invalid option. Please choose 'upgrade' or 'building'.")
+
+    # Extract errors from tar.gz, run in parallel. returs nested dict of {upgrade_id: {building_id: [error_lines]}}
+    def extract_errors_from_tar(tar_path, fails_dir, sort_order):
         job_id = re.search(r'.*simulations_job(\d+).tar.gz', tar_path).group(1)
-        job_fails_name = f'job{job_id}_failures.log'
+        job_fails_name = f'job{job_id}_failures_{sort_order}.log'
         job_fails_path = os.path.join(fails_dir, job_fails_name)
-        if os.path.exists(job_fails_path):
-            print(f'Already extracted failures from job {job_id} to {job_fails_path}')
-        else:
-            print(f'Extracting failures from job {job_id} to {job_fails_path}')
-            with open(job_fails_path, 'w') as job_fails:
-                job_fails.write(f'Errors from job_id={job_id}\n')
-                with tarfile.open(tar_path) as tar:
-                    for tar_member in tar.getmembers():
-                        if not 'failed.job' in tar_member.name:
-                            continue
-                        bldg_id = re.search(r'.*bldg(\d+).*', tar_member.name).group(1)
-                        up_id = re.search(r'.*up(\d+).*', tar_member.name).group(1)
-                        sing_out_name = f'./up{up_id}/bldg{bldg_id}/openstudio_output.log'
-                        job_fails.write(f'Errors from up{up_id}/bldg{bldg_id}\n')
-                        try:
-                            sing_out_bytes=tar.extractfile(sing_out_name).read()
-                            for l in sing_out_bytes.decode().split('\n'):
-                                if re.search(r'\[.* ERROR\]|\[.*<Error>.*\]', l):
-                                    job_fails.write(f'{l}\n')
-                        except KeyError as err:
-                            job_fails.write(f'Did not find openstudio_output.log for up{up_id}/bldg{bldg_id}, cannot extract failure details\n')
-                    tar.close()
-        return
+
+        failed_jobs = []
+        print(f'Extracting failures from job {job_id} to {job_fails_path}')
+        with tarfile.open(tar_path) as tar:
+            for tar_member in tar.getmembers():
+                if 'failed.job' in tar_member.name:
+                    bldg_id = re.search(r'.*bldg(\d+).*', tar_member.name).group(1)
+                    up_id = re.search(r'.*up(\d+).*', tar_member.name).group(1)
+                    if [up_id,bldg_id] not in failed_jobs:
```

Comment:
If we use `set()` for failed_jobs, the [up_id, bldg_id], the `in` calling could save sometime. doc of sets: https://docs.python.org/3/tutorial/datastructures.html#sets